### PR TITLE
New edge_softmax op 

### PR DESCRIPTION
**Pull request by ranzhejiang**
## Description

Fuse edge_softmax op in cpu. We obeserve that  the acceleration effect is better as CPU cores increases. The gpu op of edge_softmax will be implemented in the future
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by ranzhejiang** [comment]
## ogbn-arxiv
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: |
| 100    |  1.36   | 1.34  |    0.71  | 1.13  |
| 16    | 0.2758   |  0.1796  |    0.1464  | 0.1466  |
| 8    | 0.1428   |  0.1003  |    0.0740  | 0.0775  |
| 4    | 0.0775   |  0.0592  |    0.0407  | 0.0460  |
| 2    | 0.0348   |  0.0292  |    0.0193  | 0.0240  |
| 1    | 0.0229   |  0.0161  |    0.0114  | 0.0112  |
## cora
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: |
| 16    | 0.00183   |  0.00082  |    0.00098  | 0.00055  |
| 8    | 0.00107   |  0.00046  |    0.00056  | 0.00027  |
| 4    | 0.00091   |  0.00029  |    0.00050  | 0.00020  |
| 2    | 0.00080   |  0.00020  |    0.00047  | 0.00015  |
| 1    | 0.00075   |  0.00016  |    0.00046  | 0.00013  |

## pubmed
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: |
| 16    | 0.00623   |  0.00695  |    0.00359  | 0.00331  |
| 8    | 0.00341   |  0.00371  |    0.00172  | 0.00166  |
| 4    | 0.00238   |  0.00200  |    0.00118  | 0.00931  |
| 2    | 0.00215   |  0.00117  |    0.00104  | 0.00071  |
| 1    | 0.00136   |  0.00078  |    0.00068  | 0.00034  |

## reddit
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: |
| 8    |  4.51  |  4.99  |    2.50  |  2.91 |
| 4    | 2.23   |  2.40  |    1.284  |1.274  |
| 2    | 1.26   |  1.20  |   0.731   |  0.597 |
| 1    | 0.793   |  0.622  |    0.46  |  0.294 |

**Comment by ranzhejiang** [comment]
[](https://maxiang.io/#preview)

**Comment by ranzhejiang** [comment]
[https://maxiang.io/#preview](url)

**Comment by ranzhejiang** [comment]
image
image
image
image

**Comment by ranzhejiang** [comment]
# g3.8xlarge (32 vcpus)
## ogbn-arxiv
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: | :------: |
| 16    | 0.1004   |  0.0593  |    0.0558  | 0.0445  |
| 8    | 0.0515   |  0.0327  |    0.0289  | 0.0250  |
| 4    | 0.0267   |  0.0178  |    0.0151  | 0.0141  |
| 2    | 0.0103   |  0.0094  |    0.0069  | 0.0071  |
| 1    | 0.0073   |  0.0060  |    0.0044  | 0.0043  |
## cora
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: | :------: |
| 16    | 0.00248   |  0.00035  |    0.00054  | 0.00026  |
| 8    | 0.00082   |  0.00021  |    0.00041  | 0.00018  |
| 4    | 0.00071   |  0.00015  |    0.00037  | 0.00015  |
| 2    | 0.00061   |  0.00011  |    0.00037  | 0.00012  |
| 1    | 0.00056   |  0.000097  |    0.00035  | 0.00011  |

## pubmed
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: | :------: |
| 16    | 0.0027   |  0.0026  |    0.0014  | 0.0015  |
| 8    | 0.0020  |  0.0018  |    0.0010  | 0.0010  |
| 4    | 0.0019   |  0.0009 |    0.0009  | 0.0006  |
| 2    | 0.0021   |  0.0005  |    0.0006  | 0.0003  |
| 1    | 0.0009   |  0.0004  |    0.0005  | 0.0002  |

## reddit
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: | :------: |
| 8    |  1.631 |  1.574 |   1.008  | 0.901 |
| 4    |0.871   | 0.789  |  0.538   | 0.426|
| 2    |0.481   |  0.416  |  0.289   |  0.207 |
| 1    | 0.309   |  0.210  |    0.183  |  0.101 |


# p3.2xlarge (8 vcpus)
## ogbn-arxiv
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: |:------: |
| 16    | 0.2758   |  0.1796  |    0.1464  | 0.1466  |
| 8    | 0.1428   |  0.1003  |    0.0740  | 0.0775  |
| 4    | 0.0775   |  0.0592  |    0.0407  | 0.0460  |
| 2    | 0.0348   |  0.0292  |    0.0193  | 0.0240  |
| 1    | 0.0229   |  0.0161  |    0.0114  | 0.0112  |
## cora
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: |:------: |
| 16    | 0.00183   |  0.00082  |    0.00098  | 0.00055  |
| 8    | 0.00107   |  0.00046  |    0.00056  | 0.00027  |
| 4    | 0.00091   |  0.00029  |    0.00050  | 0.00020  |
| 2    | 0.00080   |  0.00020  |    0.00047  | 0.00015  |
| 1    | 0.00075   |  0.00016  |    0.00046  | 0.00013  |

## pubmed
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: |:------: |
| 16    | 0.00623   |  0.00695  |    0.00359  | 0.00331  |
| 8    | 0.00341   |  0.00371  |    0.00172  | 0.00166  |
| 4    | 0.00238   |  0.00200  |    0.00118  | 0.000931  |
| 2    | 0.00215   |  0.00117  |    0.00104  | 0.00071  |
| 1    | 0.00136   |  0.00078  |    0.00068  | 0.00034  |

## reddit
|   # dim    |     # origin forward edge_softmax  op |   # new forward edge_softmax op   | # origin backward edge_softmax  op |# new backward edge_softmax op|
| :--------: | :--------:| :------: | :------: |:------: |
| 8    |  4.51  |  4.99  |    2.50  |  2.91 |
| 4    | 2.23   |  2.40  |    1.284  |1.274  |
| 2    | 1.26   |  1.20  |   0.731   |  0.597 |
| 1    | 0.793   |  0.622  |    0.46  |  0.294 |

**Comment by ranzhejiang** [comment]
### End to end time with GAT model 
### Platform: g4dn.8xlarge (32 vcpus)
## reddit
|   head    |   gat with origin edge_softmax op|   gat with new edge_softmax op   | 
| :--------: | :--------:| :------: |
| 8    |  29.2 | 28.7 | 
| 4    |  20.6   |   19.5 | 
| 1    |   13.3  | 12.6  | 

## cora
|   head    |   gat with origin edge_softmax op|   gat with new edge_softmax op   | 
| :--------: | :--------:| :--------: |
| 8    | 0.0168    | 0.0156 |
| 4    |  0.0153   |  0.0143  | 
| 1    |   0.0143  | 0.0133 | 

##citeseer
|   head    |   gat with origin edge_softmax op|   gat with new edge_softmax op   | 
| :--------: | :--------:| :--------: |
| 8    | 0.0359    |0.0.0338｜
| 4    |  0.0337   |0.0322  | 
| 1    |  0.0324  |0.0307 | 

## pubmed
|   head    |   gat with origin edge_softmax op|   gat with new edge_softmax op   | 
| :--------: | :--------:| :------: |
| 8    |  0.0464 | 0.0442 | 
| 4    |  0.0405   |   0.0384 | 
| 1    |   0.0357  | 0.0349  |

**Comment by yzh119** [comment]
I think you should also compare against the performance of using edge-parallel for normalizing weights on edges, add test cases to deal with multi-dimensional edge weights.


Also,  please format the code with clang-formats.

**Comment by ranzhejiang** [comment]
> I think you should also compare against the performance of using edge-parallel for normalizing weights on edges, add test cases to deal with multi-dimensional edge weights.
> 
> Also, please format the code with clang-format.

>thank you, I'm  trying

**Comment by jermainewang** [comment]
> I think you should also compare against the performance of using edge-parallel for normalizing weights on edges, add test cases to deal with multi-dimensional edge weights.
> 
> Also, please format the code with clang-format.

How to normalize edge weights by edge parallel?

**Comment by yzh119** [comment]
I mean use node parallel for reductions and edge parallel for edge-wise computations.
I don't think it matters for edge softmax on CPU for scalar edge shape.

**Comment by ranzhejiang** [comment]
> I mean use node parallel for reductions and edge parallel for edge-wise computations. I don't think it matters for edge softmax on CPU for scalar edge weight.

I seems that using spmmcsr for  reductions to target node and sddmmcoo for edge-wise computations ?

**Comment by yzh119** [comment]
Yes I mean I'm curious how much benefit comes from fewer FFI calls.

**Comment by ranzhejiang** [comment]
> Yes I mean I'm curious how much benefit comes from fewer FFI calls.

In fact, I have implemented this cuda op for backward  and  am testing its performance, but there is a phenomenon that， I have to need temporary variable to store some result to prevent double calculation or reading，because all operation need to be done in one kernel. But the overhead may outweigh the benefit in some datasets, no matter cpu or gpu

**Comment by jermainewang** [comment]
> Yes I mean I'm curious how much benefit comes from fewer FFI calls.

I think the benefit comes more from fewer access to global memory chunks. See [Huang et. al.,](https://dl.acm.org/doi/pdf/10.1145/3437801.3441585) section 3.2 observation 3. And I think the benefit will be more significant on GPU. This PR only has CPU implementation which is perhaps why the end-to-end improvement is limited.

**Comment by fmellomascarenhas** [comment]
From an user perspective, it would be interesting to see how it performs compared to [pytorch_scatter softmax ](https://github.com/rusty1s/pytorch_scatter/blob/605566ec0c8d60867439a03cf2ba0eae75f7f532/torch_scatter/composite/softmax.py#L9)

**Comment by VoVAllen** [comment]
@dgl-bot run m6i.12xlarge edgesoftmax dmlc/master

File: python/dgl/backend/pytorch/sparse.py [hunk 1]
```diff
@@ -1,10 +1,9 @@
 import torch as th
 from distutils.version import LooseVersion
 from ...base import is_all, ALL
-from ...sparse import _gspmm, _gspmm_hetero, _gsddmm, _gsddmm_hetero, _segment_reduce, _bwd_segment_cmp
+from ...sparse import _gspmm, _gspmm_hetero, _gsddmm, _gsddmm_hetero, _segment_reduce, _bwd_segment_cmp, _edge_softmax_forward,_edge_softmax_backward
```
Comment (VoVAllen) [inline python/dgl/backend/pytorch/sparse.py:4]: Add space after comma in `_edge_softmax_forward,_edge_softmax_backward`

File: python/dgl/backend/pytorch/sparse.py [hunk 2]
```diff
@@ -1,10 +1,9 @@
...
 from ...sparse import _csrmm, _csrsum, _csrmask, _scatter_add, _update_grad_minmax_hetero
 from ...heterograph_index import create_unitgraph_from_csr
-
```
Comment (VoVAllen) [inline python/dgl/backend/pytorch/sparse.py:7]: Don't remove this blank line

File: python/dgl/backend/pytorch/sparse.py [hunk 3]
```diff
@@ -470,10 +469,13 @@ def forward(ctx, gidx, score, eids, norm_by):
             gidx = gidx.edge_subgraph([eids], True).graph
         if norm_by == 'src':
             gidx = gidx.reverse()
-        score_max = _gspmm(gidx, 'copy_rhs', 'max', None, score)[0]
-        score = th.exp(_gsddmm(gidx, 'sub', score, score_max, 'e', 'v'))
-        score_sum = _gspmm(gidx, 'copy_rhs', 'sum', None, score)[0]
-        out = _gsddmm(gidx, 'div', score, score_sum, 'e', 'v')
+        if(score.is_cuda):
+            score_max = _gspmm(gidx, 'copy_rhs', 'max', None, score)[0]
```
Comment (VoVAllen) [inline python/dgl/backend/pytorch/sparse.py:473]: Add comment to indicate currently only cpu implementation is available.

File: src/array/cpu/spmm.cc [hunk 4]
```diff
@@ -192,6 +222,56 @@ template void SpMMCsrHetero<kDLCPU, int64_t, 64>(
     const std::vector<dgl_type_t>& ufeat_node_tids,
     const std::vector<dgl_type_t>& out_node_tids);
 
+template void Edge_softmax_csr_forward<kDLCPU, int32_t, 16>(
```
Comment (VoVAllen) [inline src/array/cpu/spmm.cc:225]: Why declaring 16bits here?

**Review by VoVAllen — COMMENTED** [review]
Overall looks good to me. Please address the comment about

File: benchmarks/benchmarks/kernel/bench_edgesoftmax.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import time
+import dgl
+import torch
+
+from .. import utils
+
+# The benchmarks for ops edge_softmax
+@utils.benchmark('time', timeout=600)
+@utils.parametrize('graph', ['ogbn-arxiv','reddit' 'cora', 'pubmed',])
```
Comment (VoVAllen) [inline benchmarks/benchmarks/kernel/bench_edgesoftmax.py:9]: add comma between cora and reddit

Comment (ranzhejiang) [reply benchmarks/benchmarks/kernel/bench_edgesoftmax.py]: Ok， I have fixed it

File: python/dgl/backend/pytorch/sparse.py (hunk 3 above)
Comment (ranzhejiang) [reply python/dgl/backend/pytorch/sparse.py:476]: Ok， I have added comment for this op

File: src/array/cpu/spmm.cc (hunk 4 above)
Comment (ranzhejiang) [reply src/array/cpu/spmm.cc:225]: I have changed the Edge_softmax_csr  declaring  location follow the format like SpMMCsr

**Review by VoVAllen — APPROVED** [review]
Thanks

**Comment by 451578928** [comment]
您好，信已收到，尽快给您回复。
